crates/subspace-networking/src/create/temporary_bans.rs

```diff
@@ -0,0 +1,104 @@
+use backoff::backoff::Backoff;
+use backoff::ExponentialBackoff;
+use libp2p::PeerId;
+use lru::LruCache;
+use std::num::NonZeroUsize;
+use std::ops::Add;
+use std::time::Instant;
+
+/// Details about temporary ban, used to track lifecycle of retries
+#[derive(Debug)]
+struct TemporaryBan {
+    backoff: ExponentialBackoff,
+    next_release: Option<Instant>,
+}
+
+impl TemporaryBan {
+    /// Create new temporary ban
+    fn new(backoff: ExponentialBackoff) -> Self {
+        let mut instance = Self {
+            backoff,
+            next_release: None,
+        };
+        instance.backoff.reset();
+        instance.next_release = instance
+            .backoff
+            .next_backoff()
+            .map(|duration| Instant::now().add(duration));
+        instance
+    }
+
+    /// Whether ban is currently active and not expired
+    fn is_active(&self) -> bool {
+        if let Some(next_release) = self.next_release {
+            next_release > Instant::now()
+        } else {
+            true
+        }
+    }
+
+    /// Extend temporary ban if it expired already expired, do nothing otherwise
+    fn try_extend(&mut self) {
+        let now = Instant::now();
+
+        if let Some(next_release) = self.next_release {
```

Comment:
This next_release is confusing TBH :-(

 I would think something like this:
1. is_active() should just return self.backoff.next_backoff().is_some()
2. try_extend() should just do self.backoff.reset() if self.backoff.next_backoff().is_none()

WDYT?

Reply:
`next_backoff()` is confusing too (and method comment isn't very helpful either, it comes from `backoff` library). It will return you a `Duration` how long you need to wait before making next attempt. So by calling it you're advancing backoff further into longer intervals. `reset()` will reset backoff to initial values, we certainly don't want to do that in `try_extend()`, we want to advance backoff and wait even longer instead.

`next_release` is an `Instant` at which temporary ban will be released and new connection attempt can be made. But it doesn't remove the entity itself, so if connection attempt fails, ban will be extended exponentially.

Reply:
yeah, this crate is weird for few reasons:
1. The retries happen outside the crate, it only does some book keeping based on when next_offset() gets called
2. It doesn't keep track of number of retries, or doesn't have a way to "pause/resume" the clock (e.g) say after the first dial failure, the 2nd dial/retry happens close to max_elapsed_time for some reason, there is not much time left

Something like this https://crates.io/crates/tokio-retry looks cleaner, but would leave it to you. This would require more book keeping, as it would need to wait for either the Success/OutGoingFailure message that would come later

Reply:
There is `retry` method in that crate and it works fine with decent API, it tracks retries and everything necessary. I'm just using underlying primitive for slightly different purpose. Note: I'm not doing traditional retries here, I'm delaying them.

Reply:
Sounds good